File: backend/SpotifyClone/playlist/models/playlist.py
from django.db import models
from user.models import User
from music.models import Song
from music.models import Genre
from cloudinary.models import CloudinaryField

# Playlist Model
class Playlist(models.Model):
    name = models.CharField(max_length=255)
    description = models.TextField(null=True, blank=True)
    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='playlists')
    songs = models.ManyToManyField(Song, through='PlaylistSong', related_name='playlists')
    cover_image = CloudinaryField('playlist_covers', null=True, blank=True)
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)
    is_public = models.BooleanField(default=False)

    # Bỏ logic xử lý slug vì không còn slug
    def __str__(self):
        return f"{self.name} - {self.user.username}"

    class Meta:
        indexes = [
            models.Index(fields=['name']),
            models.Index(fields=['created_at']),
            models.Index(fields=['is_public']),
        ]

File: frontend/src/types/playlist.ts
import { Genre, Song } from "./music";
import { User } from "./user";

// Playlist Model
export interface Playlist {
  id: number;
  name: string;
  description?: string;
  user_id: number;
  cover_image?: string; // URL from Cloudinary
  created_at: string; // ISO date string
  updated_at: string; // ISO date string
  is_public: boolean;
  slug: string;

  // Optional expanded relations
  user?: User;
  songs?: PlaylistSong[];
}

// PlaylistSong Model (Junction table)
export interface PlaylistSong {
  id: number;
  playlist_id: number;
  song_id: number;
  position: number;
  added_at: string; // ISO date string

  // Optional expanded relations
  playlist?: Playlist;
  song?: Song;
}

// Collection Model
export interface Collection {
  id: number;
  title: string;
  description?: string;
  collection_type:
    | "featured"
    | "trending"
    | "new_releases"
    | "genre_based"
    | "mood_based"
    | "seasonal";
  cover_image?: string; // URL from Cloudinary
  is_active: boolean;
  created_at: string; // ISO date string
  updated_at: string; // ISO date string
  slug: string;

  // Related fields
  genres: Genre[];

  // Optional expanded relations
  songs?: CollectionSong[];
}

// CollectionSong Model (Junction table)
export interface CollectionSong {
  id: number;
  collection_id: number;
  song_id: number;
  position: number;
  added_at: string; // ISO date string

  // Optional expanded relations
  collection?: Collection;
  song?: Song;
}
